src/server/browserContext.ts

```diff
@@ -310,6 +311,56 @@ export abstract class BrowserContext extends EventEmitter {
     }
     await this._closePromise;
   }
+
+  addVisitedOrigin(origin: string) {
+    this._origins.add(origin);
+  }
+
+  async storageState(): Promise<types.StorageState> {
+    const result: types.StorageState = {
+      cookies: (await this.cookies()).filter(c => c.value !== ''),
+      origins: []
+    };
+    if (this._origins.size)  {
+      const page = await this.newPage();
+      await page._setServerRequestInterceptor(handler => {
+        handler.fulfill({ body: '<html></html>' }).catch(() => {});
+      });
+      for (const origin of this._origins) {
+        const originStorage: types.OriginStorage = { origin, localStorage: [] };
+        result.origins.push(originStorage);
+        const frame = page.mainFrame();
+        await frame.goto(new ProgressController(), origin);
+        const storage = await frame._evaluateExpression(`({
+          localStorage: Object.keys(localStorage).map(name => ({ name, value: localStorage.getItem(name) })),
+        })`, false, undefined, 'utility');
+        originStorage.localStorage = storage.localStorage;
+      }
+      await page.close();
+    }
+    return result;
+  }
+
+  async setStorageState(state: types.SetStorageState) {
+    if (state.cookies)
+      await this.addCookies(state.cookies);
```

Comment:
Should we clear cookies before that?

Reply:
Good question, we would need to clear localstorage as well. And that way we are no longer composable...

Reply:
Removed the method!